apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SimpleSubscriber.cs

```diff
@@ -748,49 +780,39 @@ private class SingleChannel
                 // If it's unrecoverable, then a global hard stop occurs.
                 while (true)
                 {
-                    await _taskHelper.ConfigureAwait(_qLock.Locked(() => _qEvent.Task));
-                    List<string> acks = new List<string>();
-                    List<string> extends = new List<string>();
-                    lock (_qLock)
+                    // Wait until there are acks or lease-extensions to push to server.
+                    while (_qLock.Locked(() => _ackQueue.Count == 0 && _extendQueue.Count == 0))
                     {
-                        // _qEvent must be re-created before any actual work is done.
-                        // This ensures no work will be missed if a re-trigger occurs just after the _qEvent.Task await.
-                        _qEvent = new TaskCompletionSource<int>();
-                        // On hard-stop or pull failure, just exit immediately. No more extends/acks will be pushed.
-                        // If a recoverable pull failure, then extend/ack pushing will continue in a new StreamingPull.
-                        if (streamingPullToken.IsCancellationRequested)
-                        {
-                            break;
-                        }
-                        // Send acks that are waiting, but no more that _maxAckExtendCount
-                        while (_ackQueue.Count > 0 && acks.Count < _maxAckExtendCount)
+                        if (_softStopCts.IsCancellationRequested && _qLock.Locked(() => _unAckedMsgCount == 0))
                         {
-                            acks.Add(_ackQueue.Dequeue());
+                            // All acks sent and a stop has been requested, so cancel local _hardStopCts
+                            // to signal this SingleChannel has completed; then exit Pusher
+                            _hardStopCts.Cancel();
+                            return;
                         }
-                        // Send extends that are waiting, but no more acks+extends than _maxAckExtendCount
-                        while (_extendQueue.Count > 0 && (acks.Count + extends.Count) < _maxAckExtendCount)
+                        var isCancelled = await _taskHelper.ConfigureAwaitHideCancellation(() => _qEvent.WaitAsync(streamingPullToken));
+                        if (isCancelled)
                         {
-                            extends.Add(_extendQueue.Dequeue());
+                            return;
                         }
                     }
-                    // If anything to be sent, then send it now
-                    if (acks.Count > 0 || extends.Count > 0)
+                    // Get local copies of acks/extensions to send.
+                    var acks = _qLock.Locked(() => _ackQueue.Take(_maxAckExtendCount).ToList());
```

Comment:
Maybe just lock once and perform both parts in the same lock?

Reply:
Yes, done.